Checkout: Send correct tax data from one-click upsell to cart

sirbrillig opened the pull request:
## Proposed Changes

This PR modifies the one-click upsell dialog to correctly send tax data from the saved card to the shopping cart.

<img width="500" alt="Screenshot 2023-03-20 at 2 01 09 PM" src="https://user-images.githubusercontent.com/2036909/226428098-08771c40-3c68-4c59-a478-30986e94fcda.png">

Requires D105292-code

Fixes https://github.com/Automattic/payments-shilling/issues/1471

## Testing Instructions

Make sure that D105292-code is applied.

- Use an account that has at least one saved credit card that includes tax info (any card that has a recent purchase in a taxable jurisdiction should work).
- With this PR running, visit `/checkout/example.com/offer-plan-upgrade/business/12345` but replace `example.com` with your test site.
- In your browser's JS console run `localStorage.setItem('debug', 'calypso:upsell-nudge')` and reload the page to view debug info.
- Click to agree to the upsell.
- In your browser's JS console, verify that you see `updating cart with contact info and product` and that it includes some objects that contain contact information; that contact information comes from the card data and will be used to construct the tax data for the cart.
- In your browser's network devtools, go to the most recent POST request to the `/me/shopping-cart` endpoint and verify that it includes the same info in the `tax` field.

<img width="222" alt="Screenshot 2023-03-20 at 3 21 53 PM" src="https://user-images.githubusercontent.com/2036909/226462530-df551fcd-6035-43ac-bd3a-8c8d68080efc.png">

sirbrillig committed:
Add tax_location to PaymentMethod type

sirbrillig committed:
Update UpsellNudge to validate tax_location from card

sirbrillig committed:
Update UpsellNudge to set tax location using tax_location from card

sirbrillig committed:
Remove unused extractStoredCardMetaValue

sirbrillig committed:
Fix missing countries prop in UpsellNudge

sirbrillig committed:
Explicitly pass contact information to validate in UpsellNudge

sirbrillig committed:
Add additional debug info

sirbrillig committed:
Add Query component for countries to UpsellNudge

sirbrillig committed:
Fix country code detection of updateCartContactDetailsForCheckout

sirbrillig committed:
Add product to cart at same time as updating cart location

sirbrillig committed:
Clear location when closing modal also

sirbrillig commented:
Holy crap, there's a bunch of type errors in the code that TS is not picking up. Like this one: 

https://github.com/Automattic/wp-calypso/blob/e4621d3d2498140e2f3a35284954aaa5bd7ad979/client/my-sites/checkout/upsell-nudge/purchase-modal/content.tsx#L217

`product` must not be undefined but you can see from the optional chaining that it can be (and it frequently is in my testing). Something is broken with TypeScript.

sirbrillig committed:
Fix some type mistakes

sirbrillig committed:
Do not redirect when showing modal

martian77 commented:
I tested this, but I'm not 100% sure if I'm seeing the right things or not. 
- In my JS console I see `updating cart with contact info and product` rather than `updating the cart with contact info`, and there's nothing that looks like a tax object in that call. 
- But the network POST to `/me/shopping-cart` does have the tax location as described. Taxes are also shown and are correct for the card details. 
<img width="481" alt="Screenshot 2023-03-21 at 11 04 42" src="https://user-images.githubusercontent.com/7126226/226588316-2ec88a9d-9bc3-4e6f-9e06-d8af2022a86a.png">

sirbrillig commented:
> In my JS console I see updating cart with contact info and product rather than updating the cart with contact info, and there's nothing that looks like a tax object in that call.

Yes, sorry, I changed the log message to add the product info after writing the test instructions 🤦 so you're seeing the right thing, but thanks for commenting on it. That log message does not have the tax object; rather it has the contact info and vat info that will be used to construct the tax object. I've updated the instructions to be more clear but it sounds like this is working!
